### ISSUE TITLE
Update stateful events data models

### ISSUE THREAD

**Issue by GGP1** [wazuh/wazuh#26568]
## Description

The stateful events [classes](https://github.com/wazuh/wazuh/blob/master/framework/wazuh/core/indexer/models/events.py) were added before the indexer templates were defined and are now outdated.

The ECS templates have suffered many changes since our modifications, including indices names, fields and types. The new templates can be found at https://github.com/wazuh/wazuh-indexer/tree/master/ecs and we should update our data models to conform with them.

**Comment by GGP1** [comment]
## Update

Updated the stateful events models and their index names. 

> [!Note]
> The issue state should be updated to `Pending review` once the triage is done.

**Comment by GGP1** [comment]
## Update

I've modified the events data model and the batcher logic to get the index depending on the event's `module` field. I have to discuss the changes with the agent team and test them before leaving it ready to review.

**Comment by GGP1** [comment]
## Update

I've been doing some changes according to the new definitions and requirements for the `POST /events/stateful` endpoint.

Regarding how to obtain the agent's metadata, instead of including it inside each event, I think the best thing to do would be to use the HTTP headers (which are going to be sent anyway) to get some of the information.

For example, the `User-Agent` header would look like this:

```
User-Agent: <name> <type> <version>
```

I wouldn't include the ID or the groups in that header, we can get the ID from the token and for the groups I would create a new header `Agent-Groups` that contains a list of groups separated by commas.

```
Agent-Groups: default,group1,group2
```

> [!Note]
> The groups could be added to the token, but during the time that the token is valid, if there are changes in the groups, they won't be taken into account.

This way, we avoid repeating information in the same request, we separate data from metadata and we have all the agent information for the type of events that require a higher number of fields (vulnerability).

Leaving the issue on hold until we have a confirmation that we can continue with this design.

**Comment by GGP1** [comment]
## Update

We met with the agent team and decided how the Comms API will receive stateful events.

### Request body

A request to the `POST /events/stateful` endpoint will have the following structure:

```json
{
  "events": [
    { // Event 1
      "module": {},
      "data": {}
    },
    { // Event 2
      "module": {},
      "data": {}
    },
    { // Event n
      "module": {},
      "data": {}
    }
  ]
}
```

Where `module` is an object that has two fields:

- **module.name**: Name of the module. One of:
  - command
  - fim
  - inventory
  - vulnerability
  - sca
- **module.type**: Data type, depending on the module name (optional). 
For `inventory`, for example:
  - network
  - package
  - process
  - system
  
> [!Note]
> Other module types may be added as new indices are created (e.g. `port`, `hotfix`, etc.)

And `data` is an object that holds the fields and values of the event itself and depends on the related [index ECS template definition](https://github.com/wazuh/wazuh-indexer/issues/270).

### Events structure

Here are some sample events for each module name and type

<details><summary>FIM</summary>

```json
{
  "module": {
    "name": "fim"
  },
  "data": {
    "file": {
      "attributes": ["string"],
      "name": "string",
      "path": "string",
      "gid": 0,
      "group": "string",
      "inode": "string",
      "mtime": "2024-10-28T18:26:10.634Z",
      "mode": "string",
      "size": 0,
      "target_path": "string",
      "type": "string",
      "uid": 0,
      "owner": "string",
      "hash": {
        "md5": "string",
        "sha1": "string",
        "sha256": "string"
      }
    },
    "registry": {
      "key": "string",
      "value": "string"
    }
  }
}
```

</details>

<details><summary>SCA</summary>

```json
{
  "module": {
    "name": "sca"
  },
  "data": {
    // To be defined, the index has no template yet
  }
}
```

</details>

<details><summary>Inventory network</summary>

```json
{
  "module": {
    "name": "inventory",
    "type": "network"
  },
  "data": {
    // To be defined, the index has no template yet
  }
}
```

</details>

<details><summary>Inventory packages</summary>

```json
{
  "module": {
    "name": "inventory",
    "type": "package"
  },
  "data": {
    "scan_time": "2024-10-28T18:26:10.634Z",
    "package": {
      "architecture": "string",
      "description": "string",
      "installed": "2024-10-28T18:26:10.634Z",
      "name": "string",
      "path": "string",
      "size": 0,
      "type": "string",
      "version": "string"
    }
  }
}
```

</details>

<details><summary>Inventory processes</summary>

```json
{
  "module": {
    "name": "inventory",
    "type": "process"
  },
  "data": {
    "scan_time": "2024-10-28T18:26:10.634Z",
    "process": {
      "pid": 0,
      "name": "string",
      "parent": {
        "pid": 0
      },
      "command_line": "string",
      "args": ["string"],
      "user": {
        "id": "string"
      },
      "real_user": {
        "id": "string"
      },
      "saved_user": {
        "id": "string"
      },
      "group": {
        "id": "string"
      },
      "real_group": {
        "id": "string"
      },
      "saved_group": {
        "id": "string"
      },
      "start": "2024-10-28T18:26:10.635Z",
      "thread": {
        "id": "string"
      }
    }
  }
}
```

</details>

<details><summary>Inventory system</summary>

```json
{
  "module": {
    "name": "inventory",
    "type": "system"
  },
  "data": {
    "scan_time": "2024-10-28T18:26:10.635Z",
    "host": {
      "architecture": "string",
      "hostname": "string",
      "os": {
        "kernel": "string",
        "full": "string",
        "name": "string",
        "platform": "string",
        "version": "string",
        "type": "string"
      }
    }
  }
}
```

</details>

<details><summary>Vulnerability</summary>

```json
{
  "module": {
    "name": "vulnerability"
  },
  "data": {
    "host": {
      "os": {
        "kernel": "string",
        "full": "string",
        "name": "string",
        "platform": "string",
        "version": "string",
        "type": "string"
      }
    },
    "package": {
      "architecture": "string",
      "build_version": "string",
      "checksum": "string",
      "description": "string",
      "install_scope": "string",
      "installed": "2024-10-28T18:26:10.635Z",
      "license": "string",
      "name": "string",
      "path": "string",
      "reference": "string",
      "size": 0,
      "type": "string",
      "version": "string"
    },
    "scanner": {
      "source": "string",
      "vendor": "string"
    },
    "score": {
      "base": 0,
      "environmental": 0,
      "temporal": 0,
      "version": "string"
    },
    "category": "string",
    "classification": "string",
    "description": "string",
    "detected_at": "2024-10-28T18:26:10.635Z",
    "enumeration": "string",
    "id": "string",
    "published_at": "2024-10-28T18:26:10.635Z",
    "reference": "string",
    "report_id": "string",
    "severity": "string",
    "under_evaluation": true
  }
}
```

</details>

<details><summary>Command result</summary>

```json
{
  "module": {
    "name": "command"
  },
  "data": {
    "document_id": "string",
    "result": {
        "code": "string",
        "message": "string",
        "data": "string"
    }
  }
}
```

</details>

### Agent metadata

Each event will contain agent metadata that will be stored along it in the indexer. This metadata won't be included in the request body, but will be provided inside HTTP headers.

The agent information required based on the events indices templates is:

- **id**
- **name**
- **type**
- **version**
- **groups**

The **id** is already stored in the JWT token, so we will obtain it from there.

The **name**, **type** and **version** will be inside the `User-Agent` HTTP header, with the following format:

```
User-Agent: <name> <type> <version>
```

Similarly, the agent groups will be stored in the HTTP header `Agent-Groups` as a string, separating values with commas:

```
 Agent-Groups: default,group1,group2
```

If we were to store more information on each event - like the agent `OS` and `IP` addresses - we could either extend the `User-Agent` values or define new HTTP headers for that information.
